vm/src/pyobjectrc.rs

```diff
@@ -0,0 +1,313 @@
+use std::borrow;
+use std::fmt;
+use std::ops::Deref;
+
+use crate::common::rc::{PyRc, PyWeak};
+use crate::pyobject::{IdProtocol, PyObject, PyObjectPayload, PyRef, PyValue, TypeProtocol};
+use crate::vm::VirtualMachine;
+
+pub struct PyObjectRc<T = dyn PyObjectPayload>
+where
+    T: ?Sized + PyObjectPayload,
+    PyRc<PyObject<T>>: AsPyObjectRef,
+{
+    inner: PyRc<PyObject<T>>,
+}
+
+pub type PyObjectWeak<T = dyn PyObjectPayload> = PyWeak<PyObject<T>>;
+
+pub trait AsPyObjectRef {
+    fn _as_ref(self) -> PyRc<PyObject<dyn PyObjectPayload>>;
+}
+
+impl<T> AsPyObjectRef for PyRc<PyObject<T>>
+where
+    T: PyObjectPayload,
+{
+    fn _as_ref(self) -> PyRc<PyObject<dyn PyObjectPayload>> {
+        self
+    }
+}
+
+impl AsPyObjectRef for PyRc<PyObject<dyn PyObjectPayload>> {
+    fn _as_ref(self) -> PyRc<PyObject<dyn PyObjectPayload>> {
+        self
+    }
+}
+
+impl<T> PyObjectRc<T>
+where
+    T: ?Sized + PyObjectPayload,
+    PyRc<PyObject<T>>: AsPyObjectRef,
+{
+    pub fn into_raw(this: Self) -> *const PyObject<T> {
+        let ptr = PyRc::as_ptr(&this.inner);
+        std::mem::forget(this);
+        ptr
+    }
+
+    unsafe fn into_rc(this: Self) -> PyRc<PyObject<T>> {
+        std::mem::transmute(this)
```

Comment:
Any reason you can't just do `this.inner` ?

Reply:
Then it will try to move inner as return value, which means `this` will have undefined inner field.

Reply:
Huh, TIL [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=311b49495470d90fb0a052649c5a600c) is an error. Maybe something like the into_raw method, but you `ptr::read(&this.inner)` + `mem::forget(this)`?

Reply:
Ok, it seems `PyRc::from_raw(Self::into_raw(this))` works